### PR TITLE
Performance update part 1

### PR DESCRIPTION
**Pull request by fanantoxa**
I've finished fir performance update.
Benchmarks from master:
```
@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@

Benchmark with :nokogiri parser

@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@

       user     system      total        real
for_tag  8.020000   0.010000   8.030000 (  8.032753)
at_depth  6.150000   0.000000   6.150000 (  6.147006)
within  7.890000   0.020000   7.910000 (  7.912147)
composite  5.480000   0.130000   5.610000 (  5.609364)

##########################################################

for_tag:   100000 artist elements parsed
at_depth:  400000 elements parsed
within:    100000 artists children parsed
composite: 0 names within artist nested 3 tags deep parsed

@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@

Benchmark with :ox parser

@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@

       user     system      total        real
for_tag  5.390000   0.110000   5.500000 (  5.512138)
at_depth  3.830000   0.070000   3.900000 (  3.899632)
within  5.210000   0.130000   5.340000 (  5.345308)
composite  3.030000   0.080000   3.110000 (  3.117383)

##########################################################

for_tag:   100000 artist elements parsed
at_depth:  400000 elements parsed
within:    100000 artists children parsed
composite: 0 names within artist nested 3 tags deep parsed
```

Benchmarks with changes on this PR:
```
@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@

Benchmark with :nokogiri parser

@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@

       user     system      total        real
for_tag  7.730000   0.020000   7.750000 (  7.750011)
at_depth  5.950000   0.010000   5.960000 (  5.950748)
within  7.450000   0.010000   7.460000 (  7.456025)
composite  5.480000   0.000000   5.480000 (  5.489144)

##########################################################

for_tag:   100000 artist elements parsed
at_depth:  400000 elements parsed
within:    100000 artists children parsed
composite: 0 names within artist nested 3 tags deep parsed

@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@

Benchmark with :ox parser

@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@

       user     system      total        real
for_tag  5.020000   0.010000   5.030000 (  5.031486)
at_depth  3.580000   0.010000   3.590000 (  3.583538)
within  4.930000   0.010000   4.940000 (  4.940562)
composite  3.040000   0.000000   3.040000 (  3.046136)

##########################################################

for_tag:   100000 artist elements parsed
at_depth:  400000 elements parsed
within:    100000 artists children parsed
composite: 0 names within artist nested 3 tags deep parsed
```

@soulcutter I've got some performance increase, but I'm not sure that my environment was good enough prepared for benchmark. Please checn it on you PC. And if works faster - please merge it.

### CODE REVIEW

File: lib/saxerator.rb [hunk 1]
```diff
@@ -4,6 +4,9 @@
 require 'saxerator/dsl'
 require 'saxerator/full_document'
 require 'saxerator/document_fragment'
+
+require 'saxerator/adapters/nokogiri'
```
Comment (soulcutter) [inline lib/saxerator.rb]: We can't require these without making both `nokogiri` and `ox` dependencies. The idea of excluding them here and requiring them inside the config is to make them optional so you can choose whatever XML library you want.

File: lib/saxerator/latches/child_of.rb [hunk 2]
```diff
@@ -31,7 +31,7 @@ def increment_depth(amount)
       end
 
       def depth_within_element
-        @depths.size > 0 ? @depths.last : 0
+        @depths.length > 0 ? @depths[-1] : 0
```
Comment (soulcutter) [inline lib/saxerator/latches/child_of.rb]: I did a benchmark and `size` and `length` have the same performance.

The `-1` index is slightly faster than `.last` though

File: lib/saxerator/parser/accumulator.rb [hunk 3]
```diff
@@ -5,28 +5,27 @@ def initialize(config, block)
...
       end
 
       def start_element(name, attrs = [])
-        @stack.push @builder.new(@config, name, Hash[*attrs.flatten])
+        @stack.push @config.output_type.new(@config, name, Hash[attrs])
       end
 
       def end_element(_)
-        if @stack.size > 1
+        if @stack.length > 1
```
Comment (soulcutter) [inline lib/saxerator/parser/accumulator.rb]: this change makes no difference in my benchmarks

File: lib/saxerator.rb (hunk 1 above)
Comment (fanantoxa) [inline lib/saxerator.rb]: @soulcutter ok. I'll move it back

File: lib/saxerator/parser/accumulator.rb (hunk 3 above)
Comment (fanantoxa) [inline lib/saxerator/parser/accumulator.rb]: @soulcutter I've updated it acording to https://github.com/JuanitoFatas/fast-ruby#array

File: lib/saxerator/latches/child_of.rb (hunk 2 above)
Comment (soulcutter) [inline lib/saxerator/latches/child_of.rb]: That's not how aliases work - it's still a single method dispatch. Even that fast-ruby link says it was 1.01x slower, which I don't consider significant.

**Comment by fanantoxa** [comment]
@soulcutter Updated code with your additions. Please check.

**Comment by soulcutter** [comment]
Looks good to me.

<pre>
@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@

Benchmark with :nokogiri parser

@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@

       user     system      total        real
for_tag  6.100000   0.040000   6.140000 (  6.162233)
at_depth  4.830000   0.030000   4.860000 (  4.868037)
within  6.180000   0.040000   6.220000 (  6.222976)
composite  4.660000   0.020000   4.680000 (  4.698063)

##########################################################

for_tag:   100000 artist elements parsed
at_depth:  400000 elements parsed
within:    100000 artists children parsed
composite: 0 names within artist nested 3 tags deep parsed

@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@

Benchmark with :ox parser

@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@

       user     system      total        real
for_tag  4.000000   0.030000   4.030000 (  4.035326)
at_depth  2.940000   0.020000   2.960000 (  2.961429)
within  3.920000   0.020000   3.940000 (  3.951661)
composite  2.330000   0.020000   2.350000 (  2.352623)

##########################################################

for_tag:   100000 artist elements parsed
at_depth:  400000 elements parsed
within:    100000 artists children parsed
composite: 0 names within artist nested 3 tags deep parsed
</pre>

**Pull request merged**